src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -169,77 +229,104 @@ public static List<string> GetResponseHeaders(SafeWinHttpHandle requestHandle, u
                 }
             }
 
-            // Allocate space for the buffer.
-            int charsNeeded = (int)bytesNeeded / sizeof(char);
-            var buffer = new StringBuilder(charsNeeded, charsNeeded);
+            return bufferLength;
+        }
+
+        private unsafe static bool QueryHeaders(
+            SafeWinHttpHandle requestHandle,
+            uint infoLevel,
+            char* buffer,
+            ref int bufferLength,
+            ref uint index)
+        {
+            Debug.Assert(bufferLength >= 0, "bufferLength must not be negative.");
+
+            // Convert the char buffer length to the length in bytes.
+            uint bufferLengthInBytes = (uint)bufferLength * sizeof(char);
 
-            results = Interop.WinHttp.WinHttpQueryHeaders(
+            // The WinHttpQueryHeaders buffer length is in bytes,
+            // but the API actually returns Unicode characters.
+            bool result = Interop.WinHttp.WinHttpQueryHeaders(
                 requestHandle,
                 infoLevel,
                 Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
-                buffer,
-                ref bytesNeeded,
+                new IntPtr(buffer),
+                ref bufferLengthInBytes,
                 ref index);
-            if (!results)
+
+            // Convert the byte buffer length back to the length in chars.
+            bufferLength = (int)bufferLengthInBytes / sizeof(char);
+
+            return result;
+        }
+
+        private static bool TryGetKnownReasonPhrase(
+            HttpStatusCode statusCode,
+            char[] array,
+            int startIndex,
+            int length,
+            out string reasonPhrase)
+        {
+            CharArrayHelpers.DebugAssertArrayInputs(array, startIndex, length);
+
+            reasonPhrase = HttpStatusDescription.Get(statusCode);
+
+            if (reasonPhrase == null)
             {
-                WinHttpException.ThrowExceptionUsingLastError();
+                return false;
+            }
+
+            if (CharArrayHelpers.EqualsOrdinal(reasonPhrase, array, startIndex, length))
+            {
+                return true;
             }
 
-            return buffer.ToString();
+            reasonPhrase = null;
+            return false;
```

Comment:
Same question here as on the other Try*s... won't the caller always immediately want to reify the value into a string if we couldn't get a cached one?  Why make this a Try then, rather than just returning the allocated string?